virtual/environment.go

```diff
@@ -923,3 +1003,11 @@ func formatActorCacheKey(
 	dst = append(dst, []byte(actorID)...)
 	return dst
 }
+
+// removeItem removes the item at the specified index from the slice.
+func removeItem[T any](slice []T, index int) []T {
+	if index == len(slice)-1 {
+		return slice[:index]
+	}
+	return append(slice[:index], slice[index+1:]...)
```

Comment:
this mutates the existing slice in place. Can you allocate a new one to be super duper cautious?